netket/experimental/sampler/metropolis_pt.py

```diff
@@ -171,215 +170,215 @@ def _reset(sampler, machine, parameters: PyTree, state: MetropolisPtSamplerState
     def _sample_next(
         sampler, machine, parameters: PyTree, state: MetropolisPtSamplerState
     ):
-        new_rng, rng = jax.random.split(state.rng)
-        # def cbr(data):
-        #    new_rng, rng = data
-        #    print("sample_next newrng:\n", new_rng,  "\nand rng:\n", rng)
-        #    return new_rng
-        # new_rng = hcb.call(
-        #   cbr,
-        #   (new_rng, rng),
-        #   result_shape=jax.ShapeDtypeStruct(new_rng.shape, new_rng.dtype),
-        # )
-
-        with loops.Scope() as s:
-            s.key = rng
-            s.σ = state.σ
-            s.log_prob = sampler.machine_pow * machine.apply(parameters, state.σ).real
-            s.beta = state.beta
+        def loop_body(i, s):
+            # 1 to propagate for next iteration, 1 for uniform rng and n_chains for transition kernel
+            s["key"], key1, key2, key3, key4 = jax.random.split(s["key"], 5)
+
+            # def cbi(data):
+            #    i, beta = data
+            #    print("sweep #", i, " for beta=\n", beta)
+            #    return beta
+            #
+            # beta = hcb.call(
+            #   cbi,
+            #   (i, s["beta"]),
+            #   result_shape=jax.ShapeDtypeStruct(s["beta"].shape, s["beta"].dtype),
+            # )
+
+            beta = s["beta"]
+
+            σp, log_prob_correction = sampler.rule.transition(
+                sampler, machine, parameters, state, key1, s["σ"]
+            )
+            proposal_log_prob = sampler.machine_pow * machine.apply(parameters, σp).real
 
-            # for logging
-            s.beta_0_index = state.beta_0_index
-            s.n_accepted_per_beta = state.n_accepted_per_beta
-            s.beta_position = state.beta_position
-            s.beta_diffusion = state.beta_diffusion
-
-            for i in s.range(sampler.n_sweeps):
-                # 1 to propagate for next iteration, 1 for uniform rng and n_chains for transition kernel
-                s.key, key1, key2, key3, key4 = jax.random.split(s.key, 5)
-
-                # def cbi(data):
-                #    i, beta = data
-                #    print("sweep #", i, " for beta=\n", beta)
-                #    return beta
-                # beta = hcb.call(
-                #   cbi,
-                #   (i, s.beta),
-                #   result_shape=jax.ShapeDtypeStruct(s.beta.shape, s.beta.dtype),
-                # )
-                beta = s.beta
-
-                σp, log_prob_correction = sampler.rule.transition(
-                    sampler, machine, parameters, state, key1, s.σ
+            uniform = jax.random.uniform(key2, shape=(sampler.n_batches,))
+            if log_prob_correction is not None:
+                do_accept = uniform < jnp.exp(
```

Comment:
It seems we don't have any test involving a `log_prob_correction`. Maybe its worth adding one?

Reply:
I think so. Should we put `MetropolisHamiltonianPt` into the tests?

Reply:
I think it's fine like that. We already test the standard codepath, and MetropolisPt samplers are experimental anyway.

@femtobit ?

Reply:
Fair enough.